File: Week1 (C)/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <ctype.h>

bool checksum(long num);
string checklength(long num1);

int main(void)
{
    // Initialise boolean 
    bool boolean = false;
    
    // Input credit card number
    long number = get_long("Please inset your credit card number (exclude hypen) :");

    // Formula calculate valid for sum   
    boolean = checksum(number);
    
    // Formula to count long length
    if (boolean == true)
    {
        string test = checklength(number);
        
        printf("%s\n", test);
    }
    
    else
    {
        printf("INVALID\n");
    }
}




// Used to count sum every 2 digits
bool checksum(long num)
{
    // Initialise 
    int sum = 0;
    int sum1 = 0;
    int remain[17] = {0};
    
    // Count total sum for whole length
    for (int i = 0; num > 0 ; i++)
    {
        remain[i] = num % 10;
        num /= 10;
        
        if (i % 2 != 0)
        {
            int x, y = 0;
            
            if (remain[i] * 2 > 9)
            {
                x = (remain[i] * 2) % 10;
                y = (remain[i] * 2) / 10;
                
                sum = sum + x + y;
            }
            
            else
            {
                sum = sum + (remain[i] * 2);
            }
        }
        else 
        {
            sum1 += remain[i];
        }
        
    }
    sum = sum + sum1;
    
    // Check if sum divisible by 10
    if (sum % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}




// Used to check long length
string checklength(long num1)
{
    // Initialise
    int i = 0;
    long num_ori = num1;
    
    // Calculate word length
    do
    {
        num_ori /= 10;
        i++;
    }
    while (num_ori > 0);    
    
    // For Visa and Mastercard
    if (i == 13 || i == 16) 
    {
        long remain = 0;
        remain = num1;
        
        for (int c = 0; remain > 100; c++)
        {
            remain /= 10;
        }
        
        if (remain == 51 || remain == 52 || remain == 53 || remain == 54 || remain == 55)
        {
            return "MASTERCARD";
        }
        
        remain /= 10;
        
        if (remain == 4)
        {
            return "VISA";
        }
    }
    
    // For AMEX
    if (i == 15)
    {
        long remain1 = 0;
        remain1 = num1;
        
        for (int c = 0; remain1 > 100; c++)
        {
            remain1 = remain1 / 10;
        }
        
        if (remain1 == 34 || remain1 == 37)
        {
            return "AMEX";
        }
    }
    
    // Do not meet requirement
    return "INVALID";
}

File: pset6/credit.py
# Import library to use floor function
import math


# Check total sum of whole credit card number
def checksum(credit_card):
    # Initialise and convert credit card no into string
    sum = 0
    i = 0
    tmp = credit_card
    
    # Create loop to add up every no in credit card
    while tmp > 0:
        remain = tmp % 10
        tmp = math.floor(tmp / 10)
        
        # Card no require *2 if not divisible by 2
        if i % 2 != 0:
            remain *= 2
            
            remain_string = str(remain)
            for no in range(len(remain_string)):
                sum += int(remain_string[no])
        else:
            sum += remain
         
        i += 1   

    # Check if total sum divisible by 10, valid if can
    if sum % 10 != 0:
        return False

    return True


# Check and return the card type
def card_type(number):
    # Initialise and convert credit card no into string type
    card_length = len(str(number))
    tmp = number

    # Get the first two digit of credit card
    while tmp > 100:
        tmp /= 10

    # Ensure no decimal
    tmp = math.floor(tmp)

    # Visa and Mastercard
    if card_length == 13 or card_length == 16:
        if tmp >= 51 and tmp <= 55:
            return "MASTERCARD"
        else:
            tmp = math.floor(tmp / 10)

            if tmp == 4:
                return "VISA"

    elif card_length == 15:
        if tmp == 34 or tmp == 37:
            return "AMEX"

    return "INVALID"


# Main function
def main():
    # Taking credit card no and check for alphabet or symbol
    while True:
        try:
            number = int(input("Please inset your credit card number (exclude hypen) :"))
            break
        except:
            print("Invalid credit card number. Please try again.")

    # Verify card
    card_status = checksum(number)
    if card_status == False:
        print("INVALID")
        return

    # Return and print card type
    output = card_type(number)
    print(output)


# Call main function
main()